### PR TITLE
Solves #10740

### PR DESCRIPTION
**Pull request by maartenarnst**
@trilinos/stokhos
@etphipp
@romintomasetti 

## Motivation
This PR solves #10740.

## Testing
These changes are tested in `packages/stokhos/test/UnitTest/Stokhos_TpetraCrsMatrixMPVectorUnitTest.hpp`.

### CODE REVIEW

File: packages/stokhos/src/sacado/kokkos/vector/belos/Belos_TpetraAdapter_MP_Vector.hpp [hunk 1]
```diff
@@ -324,6 +324,44 @@ namespace Belos {
       return mv.isConstantStride ();
     }
 
+#ifdef KOKKOS_ENABLE_CUDA
+    #define COPY_2D_VIEW_WITH_FIX_NO_UVM(__viewIn__,__viewOut__)                                 \
+    if (std::is_same<typename execution_space::memory_space, Kokkos::CudaSpace>::value)          \
```
[the remark this answers is not in the document]
Comment (maartenarnst) [reply packages/stokhos/src/sacado/kokkos/vector/belos/Belos_TpetraAdapter_MP_Vector.hpp]: Thanks for the feedback!

The patch concerns deep copies between host and device views.  We found that in OpenMP and Cuda-with-UVM builds, the current code works.  But, in Cuda-without-UVM builds, we found runtime errors, saying that a deep-copy mechanism is missing.

I agree with your concern that the previous version of the patch was too tied to Cuda.  I've generalised the patch.  The new version uses `Kokkos::SpaceAccessibility` to determine wether the memory space of the device view is accessible.  If yes, the deep_copy function is called as it was the case in this piece of code before this patch.  For the case in which the memory space of the device view is not accessible, this patch introduces a helper function that creates a mirror view as an intermediate step in the copy operation.   This was coded in a macro in the previous version of this patch, but I've now coded it as a helper function.

You mentioned a concern about deprecation.  The `Kokkos::SpaceAccessibility` struct is defined in `kokkos/core/src/Kokkos_Concepts.hpp`.  In this file, at the end around line 519, the code indicates that the definition of `SpaceAccessibility` outside of the `Kokkos` namespace is deprecated.  But, in this file, there does not seem to be an indication that `Kokkos::SpaceAccessibility` inside the `Kokkos` namespace is deprecated (?).

On our side, I've tested the patch in OpenMP, Cuda-with-uvm, Cuda-without-uvm, and Rocm builds.

If you should have suggestions, don't hesitate!  I'll be happy to keep working on it.

[the remark this answers is not in the document]
Comment (etphipp) [reply packages/stokhos/src/sacado/kokkos/vector/belos/Belos_TpetraAdapter_MP_Vector.hpp]: OK, I guess they just renamed it `Kokkos::SpaceAccessibility` from `Kokkos::Impl::MemorySpaceAccess`.

**Review by etphipp — APPROVED** [review]

**Comment by maartenarnst** [comment]
Thanks, Eric, for approving this PR.  On my side, I do not have the rights to move this forward.  Would you have a moment to set any labels that may be needed to move this forward?  The bot is asking for `AT: PRE-TEST INSPECTED`.  Thanks a lot!

**Comment by maartenarnst** [comment]
Thanks, Eric!

**Comment by etphipp** [comment]
Autotester has been down for a while, which is why this took so long to merge.

**Pull request merged**